src/main/java/BallVision.java

```diff
@@ -10,107 +11,234 @@ Basic vision algorithm for detecting balls (powercells)
-        return Math.atan(((double)(val - centerVal)) / focalLen);
+        return Math.atan(((double) (val - centerVal)) / focalLen);
     }
 
-    public static Mat algorithm(Mat input) {
+    public static MatContourPair ballDetectPipeline(Mat input) {
         // Resize
         Mat resizeInput = input;
         double resizeWidth = 320.0;
-        double resizeHeight = 240.0;
+        double resizeHeight = 181.3;
         Size toScale = new Size(resizeWidth, resizeHeight);
         Mat resizedOutput = new Mat();
         Imgproc.resize(resizeInput, resizedOutput, toScale);
 
-        /* calculate camera information
-        double aspectDiag = Math.hypot(aspectH, aspectV);
-
-        double fieldViewH = Math.atan(Math.tan(diagFieldView / 2.0) * (aspectH / aspectDiag)) * 2.0;
-        double fieldViewV = Math.atan(Math.tan(diagFieldView / 2.0) * (aspectV / aspectDiag)) * 2.0;
-
-        double hFocalLen = resizedOutput.width / (2.0 * Math.tan((fieldViewH / 2.0)));
-        double vFocalLen = resizedOutput.height / (2.0 * Math.tan((fieldViewV / 2.0)));
-
-         */
         // Blur
         Mat blurInput = resizedOutput;
-		double blurRadius = 2.7;
+        double blurRadius = 2.7;//2.7;
         double kernelSize = 2 * (blurRadius + 0.5) + 1;
         Mat blurOutput = new Mat();
         Imgproc.blur(blurInput, blurOutput, new Size(kernelSize, kernelSize));
-        
+
         // Convert to HSV and Threshold
         Mat hsvThresholdInput = blurOutput;
-		double[] hsvThresholdHue = {20.0, 180.0};
-		double[] hsvThresholdSaturation = {130.0, 255.0};
+        double[] hsvThresholdHue = {20.0, 90.0};//{20.0, 180.0};
+        double[] hsvThresholdSaturation = {130.0, 255.0};
         double[] hsvThresholdValue = {0.0, 255.0};
         Mat hsvOutput = new Mat();
         Imgproc.cvtColor(hsvThresholdInput, hsvOutput, Imgproc.COLOR_BGR2HSV);
-		Core.inRange(hsvOutput, new Scalar(hsvThresholdHue[0], hsvThresholdSaturation[0], hsvThresholdValue[0]),
-            new Scalar(hsvThresholdHue[1], hsvThresholdSaturation[1], hsvThresholdValue[1]), hsvOutput);
+        Core.inRange(hsvOutput, new Scalar(hsvThresholdHue[0], hsvThresholdSaturation[0], hsvThresholdValue[0]),
+                new Scalar(hsvThresholdHue[1], hsvThresholdSaturation[1], hsvThresholdValue[1]), hsvOutput);
 
-        //Find and draw contours
+        // Find and draw contours
         Imgcodecs.imwrite("HSV Output.jpg", hsvOutput);
         Mat contourInput = hsvOutput;
         List<MatOfPoint> contours = new ArrayList<>();
         Mat hierarchy = new Mat();
         contours.clear();
         Imgproc.findContours(contourInput, contours, hierarchy, Imgproc.RETR_LIST, Imgproc.CHAIN_APPROX_SIMPLE);
-        Imgproc.drawContours(resizedOutput, contours, -1, new Scalar(0,0,255),1);
-        
-        //For each contour found, draw bounding rectangle and marker at center if over a certain size
+        Imgproc.drawContours(resizedOutput, contours, -1, new Scalar(0, 0, 255), 1);
```

Comment:
Right now drawing the contours just gets thrown away (since the image isn't sent back to the driverstation).

If you want to send the image with countours back to the driver station (ie: for testing purposes), take a look at:
https://www.chiefdelphi.com/t/2019-vision-how-to-modify-and-display-camera-image-cpp/345723

Basically, you would modify `BallVision.process` to return `matContourPair.mat` in addition to the target location. You would also modify your `BallPipeline` to get a CameraServer instance and add a new stream, and you would make `BallPipeline.process` use `stream.putFrame` to send the drawn image back to the driverstation.

This could be very helpful for testing purposes, but we would probably want to disable it in competition (b/c it would eat up bandwidth).